Switch Async HTTP-client to `httpx` (HTTP/2-support)

thomasht86 opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

- [x] Removed `@staticmethod`-decorator of `callback_docv1` (See comment on bug below)
- [x] Switched to [httpx](https://www.python-httpx.org/http2/) - Mainly to add HTTP/2 support, which presumably could improve feed performance by a lot. 
- [x]  Unit tests passing ✅ - BUT, working to quantify performance gain. 

<details><summary>⚠️ Encountered this bug when tenacity retry is applied</summary>
<p>

```python
self = <AsyncRetrying object at 0x10468fca0 (stop=<tenacity.stop.stop_after_attempt object at 0x10468fb20>, wait=<tenacity.wa...bject at 0x10468fac0>, before=<function before_nothing at 0x104463f70>, after=<function after_nothing at 0x10447b1f0>)>
retry_state = <RetryCallState 4369142976: attempt #3; slept for 3.0; last result: failed (TypeError object dict can't be used in 'await' expression)>

    def iter(self, retry_state: "RetryCallState") -> t.Union[DoAttempt, DoSleep, t.Any]:  # noqa
        fut = retry_state.outcome
        if fut is None:
            if self.before is not None:
                self.before(retry_state)
            return DoAttempt()
    
        is_explicit_retry = fut.failed and isinstance(fut.exception(), TryAgain)
        if not (is_explicit_retry or self.retry(retry_state)):
            return fut.result()
    
        if self.after is not None:
            self.after(retry_state)
    
        self.statistics["delay_since_first_attempt"] = retry_state.seconds_since_start
        if self.stop(retry_state):
            if self.retry_error_callback:
>               return self.retry_error_callback(retry_state)
E               TypeError: 'staticmethod' object is not callable

notebooks/lib/python3.9/site-packages/tenacity/__init__.py:322: TypeError
``` 

</p>
</details> 

We should make unit tests for error handling. Maybe this is part of your fix?

jobergum commented:
Last time I tried httpx with http/2 I ran into issues with that it did not handle the case when the server would shutdown the connection (we do in Vespa cloud to disallow extremely long connections)

thomasht86 commented:
> Last time I tried httpx with http/2 I ran into issues with that it did not handle the case when the server would shutdown the connection (we do in Vespa cloud to disallow extremely long connections)

That's very useful info. I will try to mock a test for that to check!

thomasht86 commented:
Closing for now, as httpx did not improve feeding performance notably, even with HTTP/2. 
Need to investigate further to justify the change.

thomasht86 commented:
Adding these for context:
https://github.com/encode/httpx/discussions/2112
https://github.com/encode/httpx/issues/3215

Got a prototype working. Get `Error: <ConnectionTerminated error_code:ErrorCodes.NO_ERROR, last_stream_id:2147483647, additional_data:677261636566756c>` for the last ~3k when running 10k POST-requests with vectors "concurrently". 
1k requests is approx 2x faster than `feed_iterable`, which is promising.  
Need to implement queue, retries and semaphore to benchmark properly.

thomasht86 commented:
Ok, think this should be good now! 

- [x] VespaAsync now uses `httpx`with http/2. 
- [x] Changed default `max_connections`-parameter to `VespaAsync.__init__()` to 1. 
- [x] Introduced `keepalive_expiry=10` to avoid connections being terminated by the server. 
- [x] Added `feed_async_iterable()`-method. ~2-5x faster than `feed_iterable()` for the test case in the notebooks, better with increasing number of docs.
- [x] Added `feed_performance_cloud.ipynb` that compares and demonstrates different ways of feeding to Vespa Cloud. 

![feed-cloud](https://github.com/user-attachments/assets/3777f920-b1d4-4ebe-b07f-bf11394f964e)

Some things that were explored:

- Using `asyncio.Queue`, following the pattern from `feed_iterable()`, but using a Semaphore and only simple batching for restricting memory usage of tasks was 2-3x faster (with the case from the notebook in the PR). Possibly due to overhead of switching between putting/getting from the queue.
- Different `max_queue_size` does not impact performance significantly.
- Increasing `max_connections` for the async client impacts the performance negatively, as expected with HTTP2. 

Should still be possible to get better performance by parallelizing, but leaving this for later.

**Tests**

- [x] Added unit test.
- [x] Added integration test in docker.
- [x] Added notebook which also acts as test.  
- ⚠️ Unit tests failing on linkcheck, will be fixed in different PR. 
- ⚠️ Integration-cloud test cancelled due to conflicting with other PR runs.

glebashnik commented on vespa/application.py:
```diff
@@ -582,6 +581,157 @@ def _handle_result_callback(
+            async with self.asyncio(connections=max_connections) as async_session:
+                semaphore = asyncio.Semaphore(max_workers)
+                tasks = []
+                for doc in iter:
```
Feeding docs one by one seems not very efficient.
Is there any way to feed them in batches?

thomasht86 replied:
Since this is async, all the tasks will be created immediately here, without waiting for the tasks execution, and the tasks will be executed concurrently (as long as a semaphore is available). 
To limit memory usage, there is a batching of tasks instead, to avoid memory issues, if the iterable contains more documents than fits into memory, for instance. 
Did that make it any clearer?

glebashnik replied:
Still every document is still one request that needs to be processed on the server.
A lot of network and processing overhead.
Usually any data store that needs to write some non-tiny amount of data prefers batch feeding.

glebashnik replied:
No batch feeding.

glebashnik reviewed (changes requested):
See the comments.

thomasht86 commented on vespa/application.py:
```diff
@@ -1277,46 +1436,45 @@ def __init__(
-                headers={"User-Agent": "pyvespa asyncio client"},
-            )
-        return self.aiohttp_session
+            sslcontext = False
```
If running pyvespa against a local Docker instance (which is quite common), plain HTTP will be used. 
Which auth_method that is used  also logged to the user when calling `get_application_status()`, so I think that should be sufficient.

thomasht86 commented on tests/integration/test_integration_docker.py:
```diff
@@ -1421,6 +1473,43 @@ def callback(response: VespaResponse, id: str):
+            namespace="retryapplication",
+        )
+
+    def test_retries_async(self):
```
👍 
I will add comments to make this more obvious.

thomasht86 commented on vespa/application.py:
```diff
@@ -582,6 +581,157 @@ def _handle_result_callback(
+        schema: Optional[str] = None,
+        namespace: Optional[str] = None,
+        callback: Optional[Callable[[VespaResponse, str], None]] = None,
+        operation_type: Optional[str] = "feed",
```
Absolutely, but want to keep the same signature as `feed_iterable()` for now.
Thinking that could be part of a bigger `mypy`- revamp. (once python 3.8 is EOL in Oct.)

thomasht86 commented on vespa/application.py:
```diff
@@ -582,6 +581,157 @@ def _handle_result_callback(
                 queue.join()
                 consumer_thread.join()
 
+    def feed_async_iterable(
```
Hmm, also keeping in mind that we want it to not diverge from `feed_iterable()`- method naming.
We could name it `feed_aiterable()`? Is that better?

glebashnik replied:
feed_async_iterable is better than feed_aiterable.

glebashnik commented on vespa/application.py:
```diff
@@ -582,6 +581,157 @@ def _handle_result_callback(
 
+    def feed_async_iterable(
+        self,
+        iter: Iterable[Dict],
```
In case of reserved words, I think it is worth the change.
Confusing for devs.

thomasht86 replied:
Hmm, since this is a public API, we would need to add deprecation-warning for a while first if we want to rename the keyword arg from `feed_iterable`. That should nevertheless be a separate PR. So, then the question is if this new method should have both the renamed arg, but also `iter` in the meanwhile, to avoid divergent naming. 

Maybe @jobergum could enlighten us a little on best practice here? 🙏

glebashnik replied:
Right, backwards compatibility for public APIs.
Separate PR makes sense.

jobergum replied:
Since this is a new method with a new signature we don't need to make the same mistake, it this works well, we can deprecate the sync version.
